components/layout_2020/flow/construct.rs

```diff
@@ -214,34 +214,120 @@ impl BlockContainer {
+        context: &'style LayoutContext,
+        info: &'style NodeAndStyleInfo<Node>,
+        propagated_text_decoration_line: TextDecorationLine,
+    ) -> Self {
+        let text_decoration_line =
+            propagated_text_decoration_line | info.style.clone_text_decoration_line();
+        BlockContainerBuilder {
+            context,
+            info,
+            block_level_boxes: Vec::new(),
+            ongoing_inline_formatting_context: InlineFormattingContext::new(
+                text_decoration_line,
+                /* has_first_formatted_line = */ true,
+                /* ends_with_whitespace */ false,
+            ),
+            ongoing_inline_boxes_stack: Vec::new(),
+            anonymous_style: None,
+            anonymous_table_content: Vec::new(),
+        }
+    }
+
+    pub(crate) fn finish(mut self) -> BlockContainer {
+        debug_assert!(self.ongoing_inline_boxes_stack.is_empty());
 
-        debug_assert!(builder.ongoing_inline_boxes_stack.is_empty());
+        self.finish_anonymous_table_if_needed();
 
-        if !builder.ongoing_inline_formatting_context.is_empty() {
-            if builder.block_level_boxes.is_empty() {
+        if !self.ongoing_inline_formatting_context.is_empty() {
+            if self.block_level_boxes.is_empty() {
                 return BlockContainer::InlineFormattingContext(
-                    builder.ongoing_inline_formatting_context,
+                    self.ongoing_inline_formatting_context,
                 );
             }
-            builder.end_ongoing_inline_formatting_context();
+            self.end_ongoing_inline_formatting_context();
         }
 
-        let block_level_boxes = if context.use_rayon {
-            builder
-                .block_level_boxes
+        let context = self.context;
+        let block_level_boxes = if self.context.use_rayon {
+            self.block_level_boxes
                 .into_par_iter()
                 .map(|block_level_job| block_level_job.finish(context))
                 .collect()
         } else {
-            builder
-                .block_level_boxes
+            self.block_level_boxes
                 .into_iter()
                 .map(|block_level_job| block_level_job.finish(context))
                 .collect()
         };
 
         BlockContainer::BlockLevelBoxes(block_level_boxes)
     }
+
+    fn finish_anonymous_table_if_needed(&mut self) {
+        if self.anonymous_table_content.is_empty() {
+            return;
+        }
+
+        // Text decorations are not propagated to atomic inline-level descendants.
+        let inline_table = !self.ongoing_inline_boxes_stack.is_empty();
+        let propagated_text_decoration_line = if inline_table {
+            TextDecorationLine::NONE
+        } else {
+            self.ongoing_inline_formatting_context.text_decoration_line
+        };
+
+        let contents: Vec<AnonymousTableContent<'dom, Node>> =
+            self.anonymous_table_content.drain(..).collect();
```

Comment:
Would `let contents = std::mem::take(&mut self.anonymous_table_content);` be more performant?
Ditto in `components/layout_2020/table/construct.rs`

Reply:
I don't think we can use `std::mem::take` here since that does a move, but maybe we could do `std::mem::replace`. I'm not sure if it's more performant, but I think it's cleaner, so I'll make that change.

Reply:
`std::mem::take` compiles fine for me, and it seems preferred: https://github.com/rust-lang/rust-clippy/issues/10818

Reply:
Didn't address this case in `flow/construct.rs`, just one in `table/construct.rs`.

There is another `drain(..)` in `table/construct.rs` but without the `collect()`, so it may be better as-is.

Reply:
> `std::mem::take` compiles fine for me, and it seems preferred: [rust-lang/rust-clippy#10818](https://github.com/rust-lang/rust-clippy/issues/10818)

> Didn't address this case in flow/construct.rs, just one in table/construct.rs.
> There is another drain(..) in table/construct.rs but without the collect(), so it may be better as-is.

Huh...I tried this, but for some reason it didn't work for me. Thanks for the fixup.